crates/pipeline-manager/src/compiler/sql_compiler.rs

```diff
@@ -0,0 +1,533 @@
+/// This is set relatively long to not flood the logs when
+/// for instance the database becomes temporarily unreachable.
+const POLL_ERROR_INTERVAL: Duration = Duration::from_millis(30_000);
+
+/// The frequency at which during SQL compilation it is checked whether
+/// the process has finished and as well whether it needs to be canceled
+/// if the program is outdated.
+const COMPILATION_CHECK_INTERVAL: Duration = Duration::from_millis(250);
+
+/// The frequency at which the deletion of SQL compilation directories of deleted pipelines occurs.
+const CLEANUP_INTERVAL: Duration = Duration::from_millis(10_000);
+
+/// SQL compilation task that wakes up periodically.
+/// Sleeps inbetween cycles which affects the response time of SQL compilation.
+/// Note that the logic in this task assumes only one is run at a time.
+/// This task cannot fail, and any internal errors are caught and written to log if need-be.
+pub async fn sql_compiler_task(
+    common_config: CommonConfig,
+    config: CompilerConfig,
+    db: Arc<Mutex<StoragePostgres>>,
+) {
+    let mut last_cleanup: Option<Instant> = None;
+    loop {
+        let mut unexpected_error = false;
+
+        // Clean up
+        if last_cleanup.is_none() || last_cleanup.is_some_and(|ts| ts.elapsed() >= CLEANUP_INTERVAL)
+        {
+            if let Err(e) = cleanup_sql_compilation(&config, db.clone()).await {
+                match e {
+                    SqlCompilationCleanupError::DBError(e) => {
+                        error!("SQL compilation cleanup failed: database error occurred: {e}");
+                    }
+                    SqlCompilationCleanupError::CommonError(e) => {
+                        error!("SQL compilation cleanup failed: filesystem error occurred: {e}");
+                    }
+                }
+                unexpected_error = true;
+            }
+            last_cleanup = Some(Instant::now());
+        }
+
+        // Compile
+        if let Err(e) =
+            attempt_end_to_end_sql_compilation(&common_config, &config, db.clone()).await
+        {
+            match e {
+                DBError::UnknownPipeline { pipeline_id } => {
+                    debug!("SQL compilation canceled: pipeline {pipeline_id} no longer exists");
+                }
+                DBError::OutdatedProgramVersion {
+                    outdated_version,
+                    latest_version,
+                } => {
+                    debug!("SQL compilation canceled: pipeline program version ({outdated_version}) is outdated by latest ({latest_version})");
+                }
+                e => {
+                    unexpected_error = true;
+                    error!("SQL compilation canceled: unexpected database error occurred: {e}");
+                }
+            }
+        }
+
+        // Wait
+        if unexpected_error {
+            sleep(POLL_ERROR_INTERVAL).await;
+        } else {
+            sleep(POLL_INTERVAL).await;
```

Comment:
IMO this should only sleep after it has determined that there isn't any new work waiting?

Reply:
Added that the compilation attempt now returns `Ok(bool)` with boolean indicating found or not, and then only sleeping very shortly if it is true.